### PR TITLE
Always clear diagnostics when ending a debug session

### PR DESCRIPTION
**Pull request by davidwengier**
Fixes (maybe only part of?) https://github.com/dotnet/roslyn/issues/62197

FYI @tmat @dibarbet 

This fixes the issue for regular C# and Razor files, when pull diagnostics is turned on, but the issue persists in Razor with pull diagnostics turned off. Not sure what that means though :)

Explanation of the scenario and fix is in the comment in the code.

### CODE REVIEW

**Comment by dibarbet** [comment]
> but the issue persists in Razor with pull diagnostics turned off. Not sure what that means though :)

weird... the lsp pull diagnostics only changes whether or not c# diagnostics are pull.  Razor should always be pull.  I think we want it to work with the LSP pull flag off since that is how it will be in 17.3

(also this may need to target the 17.3 branch)

**Comment by davidwengier** [comment]
> I think we want it to work with the LSP pull flag off since that is how it will be in 17.3

Just to be clear, there is no issue for regular C# files with the flag off, only Razor files, but based on "the lsp pull diagnostics only changes whether or not c# diagnostics are pull", which is what I thought was the case too, I'll have to double check that. Could be my experimental hive behaving weirdly

File: src/Features/Core/Portable/EditAndContinue/EditAndContinueDiagnosticUpdateSource.cs [hunk 1]
```diff
@@ -57,11 +57,15 @@ public ValueTask<ImmutableArray<DiagnosticData>> GetDiagnosticsAsync(Workspace w
         /// Clears all diagnostics reported thru this source.
         /// We do not track the particular reported diagnostics here since we can just clear all of them at once.
         /// </summary>
-        public void ClearDiagnostics()
+        public void ClearDiagnostics(bool isSessionEnding = false)
         {
             // If ClearDiagnostics is called and there weren't any diagnostics previously, then there is no point incrementing
             // our version number and potentially invalidating caches unnecessarily.
-            if (_previouslyHadDiagnostics)
+            // If the debug session is ending, however, we want to always increment otherwise we can get stuck. eg if the user
```
Comment (tmat) [inline src/Features/Core/Portable/EditAndContinue/EditAndContinueDiagnosticUpdateSource.cs:64]: >debug [](http://example.com/codeflow?start=22&length=5)

Can we always increment the version, to simplify this?

**Comment by tmat** [comment]
@dibarbet We're gonna be able to get rid of `IDiagnosticUpdateSource` once we move fully to pull diags, right?

File: src/Features/Core/Portable/EditAndContinue/EditAndContinueDiagnosticUpdateSource.cs (hunk 1 above)
Comment (davidwengier) [reply src/Features/Core/Portable/EditAndContinue/EditAndContinueDiagnosticUpdateSource.cs:64]: We can, but it felt like maybe it was a bit too aggressive. We'd be essentially invalidating the cache that the LSP handler uses on every debugger step, even if there were no document changes.

**Comment by davidwengier** [comment]
@dibarbet @tmat Anything else to do for this? Should clearing the diagnostics be more aggressive?

**Comment by davidwengier** [comment]
This is your weekly ping @tmat @dibarbet 
If pull diagnostics is going to be turned on for people, this could become more of an issue...

**Review by dibarbet — APPROVED** [review]

**Comment by davidwengier** [comment]
Talked offline with David, will merge this and follow up with more aggressive cache invalidation if we see feedback.

**Pull request merged**